src/codecs/webp/options.tsx

```diff
@@ -1,3 +1,317 @@
-import qualityOption from '../generic/quality-option';
+import { h, Component } from 'preact';
+import { bind } from '../../lib/util';
+import { EncodeOptions, WebPImageHint } from './encoder';
+import * as styles from './styles.scss';
 
-export default qualityOption();
+type Props = {
+  options: EncodeOptions,
+  onChange(newOptions: EncodeOptions): void,
+};
+
+// From kLosslessPresets in config_enc.c
+// The format is [method, quality].
+const losslessPresets:[number, number][] = [
+  [0, 0], [1, 20], [2, 25], [3, 30], [3, 50],
+  [4, 50], [4, 75], [4, 90], [5, 90], [6, 100],
+];
+const losslessPresetDefault = 6;
+
+function determineLosslessQuality(quality: number): number {
+  const index = losslessPresets.findIndex(item => item[1] === quality);
+  if (index !== -1) return index;
+  // Quality doesn't match one of the presets.
+  // This can happen when toggling 'lossless'.
+  return losslessPresetDefault;
+}
+
+export default class WebPEncoderOptions extends Component<Props, {}> {
+  @bind
+  onChange(event: Event) {
+    const form = (event.currentTarget as HTMLInputElement).closest('form') as HTMLFormElement;
+    const lossless = Number((form.lossless as HTMLInputElement).checked);
+    const losslessPresetInput = (form.lossless_preset as HTMLInputElement);
+
+    const options: EncodeOptions = {
+      // Copy over options the form doesn't care about, eg emulate_jpeg_size
+      ...this.props.options,
+      // And now stuff from the form:
+      lossless,
+      // Special-cased inputs:
+      // In lossless mode, the quality is derived from the preset.
+      quality: lossless ?
+        losslessPresets[Number(losslessPresetInput.value)][1] :
+        Number((form.quality as HTMLInputElement).value),
+      // In lossless mode, the method is derived from the preset.
+      method: lossless ?
+        losslessPresets[Number(losslessPresetInput.value)][0] :
+        Number((form.method_input as HTMLInputElement).value),
+      image_hint: (form.image_hint as HTMLInputElement).checked ?
+        WebPImageHint.WEBP_HINT_GRAPH :
+        WebPImageHint.WEBP_HINT_DEFAULT,
+      // .checked
+      exact: Number((form.exact as HTMLInputElement).checked),
```

Comment:
I know it would be a 1-line function, but to reduce visual noise may be add a `isChecked()` function?

Reply:
Like this?

```ts
function fieldCheckedAsNumber(field: any): number {
  return Number((field as HTMLInputElement).checked);
}
```

Reply:
I guess you want a similar function for `.value`?

Reply:
It abstracts away the awkward casting, so sgtm.